fgpyo/vcf/builder.py

```diff
@@ -0,0 +1,159 @@
+    header: PysamHeader
+
+    def __init__(
+        self, samples: Optional[List[str]] = None, sd: Optional[Dict[str, Dict[str, Any]]] = None
+    ) -> None:
+        """Initializes a new VariantBuilder for generating variants and VCF files.
+
+        Args:
+            samples: the name of the sample(s)
+            sd: optional sequence dictionary
+        """
+        super().__init__(samples=samples, sd=sd)
+        self.header = PysamHeader()
+        for line in VariantBuilder.build_header_string(samples=samples, sd=self.sd).split("\n"):
+            if not line.startswith("##"):
+                break
+            self.header.add_line(line)
+        if samples is not None:
+            self.header.add_samples(samples)
+
+    def close(self) -> None:
+        """Closes this builder"""
+        pass
+
+    def add(
+        self,
+        contig: Optional[str] = None,
+        pos: Optional[int] = None,
+        id: Optional[str] = None,
+        ref: Optional[str] = None,
+        alts: Optional[Iterable[str]] = None,
+        qual: Optional[int] = None,
+        filter: Optional[Iterable[str]] = None,
+        info: Optional[Dict[str, Any]] = None,
+        samples: Optional[Dict[str, Dict[str, Any]]] = None,
+        input_start_index: int = 1,
+    ) -> PysamVariant:
+        """Generates a new variant and adds it to the internal collection.
+
+        Very little validation is done with respect to INFO and FORMAT keys being defined in the
+        header.
+
+        Args:
+            contig: the chromosome name
+            pos: the position
+            id: the variant id
+            ref: the reference allele
+            alts: the list of alternate alleles, None if no alternates
+            qual: the variant quality
+            filter: the list of filters, None if no filters (ex. PASS)
+            info: the dictionary of INFO key-value pairs
+            formats: the dictionary of sample name to FORMAT key-value pairs.
+            input_start_index: the genomic start index for input data (e.g. 0 for 0-based, 1-for
+                               1-based). VCF files are 1-based, but pysam is 0-based.
+        """
+        contig, pos, id, ref, alts, qual = VariantBuilderDefaults.init(
+            self.sd, contig, pos, id, ref, alts, qual
+        )
+
+        if contig not in self.sd:
+            raise ValueError(f"Chromosome `{contig}` not in the sequence dictionary.")
+        if samples is not None and any(sample not in self.samples for sample in samples):
+            raise ValueError(
+                "Extra sample(s) given: "
+                + ", ".join(sample for sample in samples if sample not in self.samples)
+            )
+
+        alleles = (ref,) if alts is None else (ref, *alts)
+
+        if self.samples is None or len(self.samples) == 0:
+            formats = None
+        elif samples is None or len(samples) == 0:
+            formats = [{"GT": "./."} for _ in self.samples]
```

Comment:
The current behavior is:
* if the VCF is sites-only, don't supply samples/formats
* if the VCF has samples and none are provided jam in a bunch of no-call GTs
* if the VCF has samples and some are provided, set the missing samples to no-call for every field in `all_format_keys`

So I guess the intention is to distinguish a sites-only VCF (no format data) from a VCF with calls where for whatever reason we've provided no calls for a variant. My thinking is that it would be slightly preferable to
a) use the format header fields and set all of those to no-call, or b) just throw an error. Unless there's objections, I'm planning to work up option a).

Reply:
After trying a bunch of things, some more complicated than others, I settled on just setting the format values to None (i.e. provide nothing). This does not produce an error, nor does it cause `GT` to appear when it otherwise might not be added. It's easy enough to create an all no-call GT that the user should just do so if that's what they want:
```
sample_formats = {sample_id: {"GT": "."} for sample_id in builder.sample_ids})
```
and if they don't care about FORMAT values, the least surprising thing is for them to be empty. Happy to change this if there are strong opposing feelings or it becomes inconvenient down the line.